src/algebra/char_zero.lean

```diff
@@ -73,10 +73,20 @@ def cast_embedding : ℕ ↪ R := ⟨coe, cast_injective⟩
 cast_injective.eq_iff
 
 @[simp, norm_cast] theorem cast_eq_zero {n : ℕ} : (n : R) = 0 ↔ n = 0 :=
-by rw [← cast_zero, cast_inj]
+by rw [←cast_zero, cast_inj]
+
+@[simp, norm_cast] theorem cast_eq_one {n : ℕ} : (n : R) = 1 ↔ n = 1 :=
+by rw [←cast_one, cast_inj]
+
+@[simp] lemma cast_pow_eq_one {R : Type*} [semiring R] [char_zero R] (q : ℕ) (n : ℕ) (hn : n ≠ 0) :
+  (q : R) ^ n = 1 ↔ q = 1 :=
```

Comment:
Is it worth considering this alternative (I'm not sure myself):
```suggestion
@[simp] lemma cast_pow_succ_eq_one {R : Type*} [semiring R] [char_zero R] (q : ℕ) (n : ℕ) :
  (q : R) ^ (n+1) = 1 ↔ q = 1 :=
```

Reply:
I agree that this makes a better simp-lemma.
But maybe we should also keep the original version for `rw`ing.

Reply:
Does `simp` not have succ_pos/succ_ne_zero?

Reply:
hmm, `simp` just doesn't seem to like simplifying this sort of expression at all:

```lean
import algebra.char_zero
import data.rat.basic

@[simp] lemma cast_pow_succ_eq_one {R : Type*} [semiring R] [char_zero R] (q : ℕ) (n : ℕ) :
  (q : R) ^ (n+1) = 1 ↔ q = 1 :=
nat.cast_pow_eq_one _ _ n.succ_ne_zero

example {n t : ℕ} : ((n : ℚ) ^ (t + 1) = 1) ↔ false :=
begin
  simp [nat.cast_pow_eq_one _ _ t.succ_ne_zero],
end
```

Reply:
This is a red herring, the problem is a missing instance `char_zero ℚ` which will be picked up if you change:
```lean
import data.rat.basic
```
to
```lean
import data.rat.order
```

Reply:
However you're right about `simp` finding `nat.succ_ne_zero`. For example the following works:
```lean
import algebra.char_zero
import data.rat.order

variables {R : Type*} [semiring R] [char_zero R] (q : ℕ) (n : ℕ)

@[simp] lemma cast_cast_pow_eq_one (hn : n ≠ 0) : (q : R) ^ n = 1 ↔ q = 1 := sorry

-- @[simp] lemma cast_pow_succ_eq_one : (q : R) ^ (n+1) = 1 ↔ q = 1 := sorry

example {n t : ℕ} : ((n : ℚ) ^ (t + 1) = 1) ↔ n = 1 := by simp
```